src/main.rs

```diff
@@ -487,6 +617,55 @@ impl From<CheckRelease> for cargo_semver_checks::Check {
     }
 }
 
+/// Helper function to encapsulate the logic of validating that unstable options
+/// were not used without `-Z unstable-options` and issuing deprecation warnings
+/// for any stable feature flags that were explicitly specified.
+fn validate_feature_flags(config: &mut GlobalConfig, args: &SemverChecks) -> anyhow::Result<()> {
+    // needed to avoid borrow checker errors when printing with config.
+    let stable_flags: Vec<_> = config
+        .feature_flags()
+        .iter()
+        .filter(|x| x.stable)
+        .copied()
+        .collect();
+
+    for stable_flag in stable_flags {
+        config
+            .shell_warn(format_args!(
+                "the feature flag {} has been stabilized and may be removed
+            from the list of feature flags in a future release.",
+                stable_flag.id
+            ))
+            .expect("printing failed");
+    }
+
+    if !config.feature_flag_enabled(FeatureFlag::UNSTABLE_OPTIONS) {
+        let unstable_options = match &args.command {
+            Some(SemverChecksCommands::CheckRelease(cr)) => &cr.unstable_options,
+            None => &args.check_release.unstable_options,
+        };
+
+        let non_default_options = unstable_options.non_default();
+
+        if !non_default_options.is_empty() {
+            let mut message = String::from(
+                "the following options are not supported without `-Z unstable-options`:\n",
+            );
+
+            for option in non_default_options {
+                use std::fmt::Write as _;
+
+                // writes to strings are infallible.
+                let _ = writeln!(&mut message, " - `{option}`");
```

Comment:
Can a write cause a reallocation which then fails?

I'd feel better about this being an `.expect("writes to strings are infallible")` than a comment. I've been on many a wild goose chase as a result of someone using a "comment assert" instead of a real assert, so I've learned to find them suspicious by default.
```suggestion
                writeln!(&mut message, " - `{option}`").expect("writes to strings are infallible");
```

Reply:
I think (confusingly) it is an [API guarantee](https://doc.rust-lang.org/std/fmt/struct.Error.html) that writes to strings are fine, but I don't think this error check would impact performance at all.